### PR TITLE
High level motion framework

### PR DESCRIPTION
**Pull request by aftersomemath**
Action server, motion planner skeleton, test task, and test driver for the action server

### CODE REVIEW

File: scripts/motion_planner.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python
+import sys
+import rospy
+from iarc_task_action_server import IarcTaskActionServer
+
+class MotionPlanner:
+
+    def __init__(self, action_server):
+        self.action_server = action_server
```
Comment (amiller27) [inline scripts/motion_planner.py]: These should probably be marked as private

File: scripts/motion_planner.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        self.task = None
+
+    def get_velocity_command(self):
```
Comment (amiller27) [inline scripts/motion_planner.py]: This is supposed to return a velocity, is it just not implemented yet?

File: scripts/obstacle_avoider.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env python
+import rospy
+
+class ObstacleAvoider:
+
+    def get_safe_velocity(preferred_velocity):
+        print("get_safe_velocity")
```
Comment (amiller27) [inline scripts/obstacle_avoider.py]: If you're going to use this like this, you should probably import print_function so the syntax isn't confusing

File: scripts/test_task.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                rospy.loginfo("TestTask done")
+                self.success = True
+                return True
+            else:
+                rospy.loginfo("TestTask not done")
+                return False
+        # Haven't been initialized yet
+        return False
+
+    def is_aborted(self):
```
Comment (amiller27) [inline scripts/test_task.py]: Deprecated

File: scripts/test_task.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        if self.aborted:
+            rospy.loginfo("TestTask aborted")
+        return self.aborted
+
+    def get_result(self):
```
Comment (amiller27) [inline scripts/test_task.py]: Deprecated

File: action/QuadMove.action [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#goal definition
+string movement_type
+bool preempt
+float64 takeoff_height
+---
+#result definition
+bool success
```
Comment (aftersomemath) [inline action/QuadMove.action]: Why should the caller know if it was preempted? Is that adding more functionality than we need now and for a while?

Comment (amiller27) [inline action/QuadMove.action]: Just realized I didn't mean preempted, I actually meant rejected.  Although I think preemption could probably be returned as well, that could be useful for the gameplay algorithm.  Either way, I think this can stay for now and go into another PR if we want to add that functionality.

Comment (aftersomemath) [inline action/QuadMove.action]: I agree that that is a feature request.

File: CMakeLists.txt [hunk 7]
```diff
@@ -62,17 +63,17 @@ find_package(catkin REQUIRED COMPONENTS
...
+  DIRECTORY action
+  FILES QuadMove.action
+)
 
 ## Generate added messages and services with any dependencies listed here
-# generate_messages(
-#   DEPENDENCIES
-#   iarc7_msgs
-# )
+generate_messages(
```
Comment (aftersomemath) [inline CMakeLists.txt]: Yes, above the line in question a new action was added to the package.

File: scripts/abstract_task.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python
+import rospy
+
+class AbstractTask:
+
+    # Abstract method
+    def get_preferred_velocity(self):
```
Comment (aftersomemath) [inline scripts/abstract_task.py]: Passing in a timestamp means that the task has to be prepared for a reasonable range of timestamps to be entered or return an error. I'd rather have the relationship implicitly be the current time.

File: action/QuadMove.action (hunk 6 above)
Comment (amiller27) [inline action/QuadMove.action]: I made a trello card

File: scripts/abstract_task.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    # Abstract method
+    def cancel(self):
+        raise NotImplementedError("Subclass must implement abstract method")
+
+    # Abstract method
+    def is_done(self):
+        raise NotImplementedError("Subclass must implement abstract method")
+
+    # Abstract method
```
Comment (aftersomemath) [inline scripts/abstract_task.py]: Fixed

File: scripts/abstract_task.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    def is_aborted(self):
+        raise NotImplementedError("Subclass must implement abstract method")
+
+    # Abstract method
+    def get_result(self):
```
Comment (aftersomemath) [inline scripts/abstract_task.py]: Fixed

File: scripts/takeoff_task.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+from abstract_task import AbstractTask
+
+class TakeoffTask(AbstractTask):
+
+    def __init__(self, takeoff_height):
+        self.takeoff_height = takeoff_height
+
+    def get_preferred_velocity(self):
+        rospy.loginfo("get_preferred_velocity")
+        return 5
```
Comment (aftersomemath) [inline scripts/takeoff_task.py]: Fixed

File: scripts/takeoff_task.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+    def cancel(self):
+        return True
+
+    def is_done(self):
+        return True
+
+    def is_aborted(self):
```
Comment (aftersomemath) [inline scripts/takeoff_task.py]: Fixed

File: scripts/takeoff_task.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        return False
+
+    def get_result(self):
```
Comment (aftersomemath) [inline scripts/takeoff_task.py]: Fixed

File: scripts/test_task.py (hunk 4 above)
Comment (aftersomemath) [inline scripts/test_task.py]: Fixed

File: scripts/test_task.py (hunk 5 above)
Comment (aftersomemath) [inline scripts/test_task.py]: Fixed

File: scripts/iarc_task_action_server.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+from takeoff_task import TakeoffTask
+from test_task import TestTask
+from iarc7_motion.msg import QuadMoveAction, QuadMoveResult
+
+class IarcTaskActionServer:
+    def __init__(self):
+        self._action_name = "motion_planner_server"
+
+
+        self.action_server = actionlib.ActionServer(self._action_name,
```
Comment (aftersomemath) [inline scripts/iarc_task_action_server.py]: Fixed

File: scripts/motion_planner.py (hunk 1 above)
Comment (aftersomemath) [inline scripts/motion_planner.py]: Fixed

File: scripts/iarc_task_action_server.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        self.goal_tasks = []
+        self.current_task = None
+        self.current_goal = None
+        self.cancel_requested = False
+        self.lock = threading.RLock()
+        # Start action server last to avoid race condition
+        self.action_server.start()
+
+    # Private method
+    def new_goal(self, goal):
```
Comment (aftersomemath) [inline scripts/iarc_task_action_server.py]: Fixed

File: scripts/motion_planner.py (hunk 2 above)
Comment (aftersomemath) [inline scripts/motion_planner.py]: yes

File: scripts/obstacle_avoider.py (hunk 3 above)
Comment (aftersomemath) [inline scripts/obstacle_avoider.py]: fixed

File: scripts/iarc_task_action_server.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        with self.lock:
+            rospy.logdebug("new_goal: %s", goal.get_goal_id().id)
+
+            task_request = goal.get_goal()
+
+            if task_request.movement_type == "takeoff" :
```
Comment (aftersomemath) [inline scripts/iarc_task_action_server.py]: Fixed

File: scripts/iarc_task_action_server.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+                new_task = TakeoffTask(task_request.takeoff_height)
+            elif task_request.movement_type == "test_task":
+                # Uses the preempt to decide whether or not to test aborting
+                new_task = TestTask(task_request.takeoff_height)
+            else:
```
Comment (aftersomemath) [inline scripts/iarc_task_action_server.py]: Fixed

File: scripts/iarc_task_action_server.py [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+                if len(self.goal_tasks) > 0:
+                    for x, _ in  self.goal_tasks:
+                        x.set_cancel_requested()
+                        x.set_canceled()
+                    self.goal_tasks = []
+                if self.current_goal:
+                    self.cancel_requested = True
+                    self.current_goal.set_cancel_requested()
+
+            self.goal_tasks.append([goal, new_task])
```
Comment (aftersomemath) [inline scripts/iarc_task_action_server.py]: Fixed

File: scripts/iarc_task_action_server.py (hunk 16 above)
Comment (aftersomemath) [inline scripts/iarc_task_action_server.py]: This creates a side affect because initially I was thinking that what each task was passed on instantiation could be adjusted. Fixing this should be a new trello card, it requires restructuring of the action and the tasks.
